conda build has non-zero exit status despite successful build

### What happened?

When building a recipe with conda-build version 24.1.0, the return status is non-zero despite the build being successful. This does not happen with conda-build version 3.28.4.

### Conda Details

<details>
<summary><code>conda info</code></summary>

```shell
     active environment : /…/envs/miniconda3
    active env location : /…/envs/miniconda3
            shell level : 2
       user config file : /…/.condarc
 populated config files : /…/.condarc
          conda version : 23.11.0
    conda-build version : 24.1.0
         python version : 3.11.5.final.0
                 solver : libmamba (default)
       virtual packages : __archspec=1=zen
                          __conda=23.11.0=0
                          __cuda=12.3=0
                          __glibc=2.38=0
                          __linux=6.6.13=0
                          __unix=0=0
       base environment : /…/miniconda3  (writable)
      conda av data dir : /…/miniconda3/etc/conda
  conda av metadata url : None
           channel URLs : https://conda.anaconda.org/sebp/linux-64
                          https://conda.anaconda.org/sebp/noarch
                          https://repo.anaconda.com/pkgs/main/linux-64
                          https://repo.anaconda.com/pkgs/main/noarch
                          https://repo.anaconda.com/pkgs/r/linux-64
                          https://repo.anaconda.com/pkgs/r/noarch
          package cache : /…/miniconda3/pkgs
                          /…/.conda/pkgs
       envs directories : /…/miniconda3/envs
                          /…/.conda/envs
               platform : linux-64
             user-agent : conda/23.11.0 requests/2.31.0 CPython/3.11.5 Linux/6.6.13-200.fc39.x86_64 fedora/39 glibc/2.38 solver/libmamba conda-libmamba-solver/23.12.0 libmambapy/1.5.3 aau/0.4.3 c/K7pj3PuF7vhwR8S115QX0A s/oxhLIYnOkyClhlSbRCzyKA e/oS9vSrZdGoblDAkkam1dig
                UID:GID : 1000:1000
             netrc file : None
           offline mode : False
```

</details>

<details>
<summary><code>conda config</code></summary>

```shell
add_pip_as_python_dependency: False
auto_update_conda: False
changeps1: False
channel_priority: flexible
channels:
  - sebp
  - defaults
show_channel_urls: True
always_yes: True
report_errors: False
anaconda_upload: False
conda_build:
  pkg_format: 2
```

</details>

<details>
<summary><code>conda list</code></summary>

```shell
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main    defaults
_openmp_mutex             5.1                       1_gnu    defaults
anaconda-anon-usage       0.4.3           py311hfc0e8ea_100    defaults
anaconda-client           1.12.2          py311h06a4308_0    defaults
archspec                  0.2.1              pyhd3eb1b0_0    defaults
attrs                     23.1.0          py311h06a4308_0    defaults
beautifulsoup4            4.12.2          py311h06a4308_0    defaults
boltons                   23.0.0          py311h06a4308_0    defaults
brotli-python             1.0.9           py311h6a678d5_7    defaults
bzip2                     1.0.8                h7b6447c_0    defaults
c-ares                    1.19.1               h5eee18b_0    defaults
ca-certificates           2023.12.12           h06a4308_0    defaults
certifi                   2023.11.17      py311h06a4308_0    defaults
cffi                      1.16.0          py311h5eee18b_0    defaults
chardet                   4.0.0           py311h06a4308_1003    defaults
charset-normalizer        2.0.4              pyhd3eb1b0_0    defaults
click                     8.1.7           py311h06a4308_0    defaults
conda                     23.11.0         py311h06a4308_0    defaults
conda-build               24.1.0          py311h06a4308_0    defaults
conda-content-trust       0.2.0           py311h06a4308_0    defaults
conda-index               0.4.0              pyhd3eb1b0_0    defaults
conda-libmamba-solver     23.12.0            pyhd3eb1b0_1    defaults
conda-package-handling    2.2.0           py311h06a4308_0    defaults
conda-package-streaming   0.9.0           py311h06a4308_0    defaults
cryptography              41.0.7          py311hdda0065_0    defaults
defusedxml                0.7.1              pyhd3eb1b0_0    defaults
distro                    1.8.0           py311h06a4308_0    defaults
filelock                  3.13.1          py311h06a4308_0    defaults
fmt                       9.1.0                hdb19cb5_0    defaults
icu                       73.1                 h6a678d5_0    defaults
idna                      3.4             py311h06a4308_0    defaults
jinja2                    3.1.3           py311h06a4308_0    defaults
jsonpatch                 1.32               pyhd3eb1b0_0    defaults
jsonpointer               2.1                pyhd3eb1b0_0    defaults
jsonschema                4.19.2          py311h06a4308_0    defaults
jsonschema-specifications 2023.7.1        py311h06a4308_0    defaults
jupyter_core              5.5.0           py311h06a4308_0    defaults
krb5                      1.20.1               h143b758_1    defaults
ld_impl_linux-64          2.38                 h1181459_1    defaults
libarchive                3.6.2                h6ac8c49_2    defaults
libcurl                   8.4.0                h251f7ec_1    defaults
libedit                   3.1.20230828         h5eee18b_0    defaults
libev                     4.33                 h7f8727e_1    defaults
libffi                    3.4.4                h6a678d5_0    defaults
libgcc-ng                 11.2.0               h1234567_1    defaults
libgomp                   11.2.0               h1234567_1    defaults
liblief                   0.12.3               h6a678d5_0    defaults
libmamba                  1.5.3                haf1ee3a_0    defaults
libmambapy                1.5.3           py311h2dafd23_0    defaults
libnghttp2                1.57.0               h2d74bed_0    defaults
libsolv                   0.7.24               he621ea3_0    defaults
libssh2                   1.10.0               hdbd6064_2    defaults
libstdcxx-ng              11.2.0               h1234567_1    defaults
libuuid                   1.41.5               h5eee18b_0    defaults
libxml2                   2.10.4               hf1b16e4_1    defaults
lz4-c                     1.9.4                h6a678d5_0    defaults
markupsafe                2.1.3           py311h5eee18b_0    defaults
menuinst                  2.0.1           py311h06a4308_1    defaults
more-itertools            10.1.0          py311h06a4308_0    defaults
nbformat                  5.9.2           py311h06a4308_0    defaults
ncurses                   6.4                  h6a678d5_0    defaults
openssl                   3.0.13               h7f8727e_0    defaults
packaging                 23.1            py311h06a4308_0    defaults
patch                     2.7.6             h7b6447c_1001    defaults
patchelf                  0.17.2               h6a678d5_0    defaults
pcre2                     10.42                hebb0a14_0    defaults
pip                       23.3.1          py311h06a4308_0    defaults
pkginfo                   1.9.6           py311h06a4308_0    defaults
platformdirs              3.10.0          py311h06a4308_0    defaults
pluggy                    1.0.0           py311h06a4308_1    defaults
psutil                    5.9.0           py311h5eee18b_0    defaults
py-lief                   0.12.3          py311h6a678d5_0    defaults
pybind11-abi              4                    hd3eb1b0_1    defaults
pycosat                   0.6.6           py311h5eee18b_0    defaults
pycparser                 2.21               pyhd3eb1b0_0    defaults
pyopenssl                 23.2.0          py311h06a4308_0    defaults
pysocks                   1.7.1           py311h06a4308_0    defaults
python                    3.11.5               h955ad1f_0    defaults
python-dateutil           2.8.2              pyhd3eb1b0_0    defaults
python-fastjsonschema     2.16.2          py311h06a4308_0    defaults
python-libarchive-c       2.9                pyhd3eb1b0_1    defaults
pytz                      2023.3.post1    py311h06a4308_0    defaults
pyyaml                    6.0.1           py311h5eee18b_0    defaults
readline                  8.2                  h5eee18b_0    defaults
referencing               0.30.2          py311h06a4308_0    defaults
reproc                    14.2.4               h295c915_1    defaults
reproc-cpp                14.2.4               h295c915_1    defaults
requests                  2.31.0          py311h06a4308_0    defaults
requests-toolbelt         1.0.0           py311h06a4308_0    defaults
rpds-py                   0.10.6          py311hb02cf49_0    defaults
ruamel.yaml               0.17.21         py311h5eee18b_0    defaults
setuptools                68.2.2          py311h06a4308_0    defaults
six                       1.16.0             pyhd3eb1b0_1    defaults
soupsieve                 2.5             py311h06a4308_0    defaults
sqlite                    3.41.2               h5eee18b_0    defaults
tk                        8.6.12               h1ccaba5_0    defaults
tqdm                      4.65.0          py311h92b7b1e_0    defaults
traitlets                 5.7.1           py311h06a4308_0    defaults
truststore                0.8.0           py311h06a4308_0    defaults
tzdata                    2023c                h04d1e81_0    defaults
urllib3                   1.26.18         py311h06a4308_0    defaults
wheel                     0.41.2          py311h06a4308_0    defaults
xz                        5.4.5                h5eee18b_0    defaults
yaml                      0.2.5                h7b6447c_0    defaults
yaml-cpp                  0.8.0                h6a678d5_0    defaults
zlib                      1.2.13               h5eee18b_0    defaults
zstandard                 0.19.0          py311h5eee18b_0    defaults
zstd                      1.5.5                hc292b87_0    defaults
```

</details>


### Additional Context

Example recipe:

```yaml
{% set name = "tqdm" %}
{% set version = "4.66.1" %}
{% set sha256 = "d88e651f9db8d8551a62556d3cff9e3034274ca5d66e93197cf2490e2dcb69c7" %}

package:
  name: {{ name|lower }}
  version: {{ version }}

source:
  fn: {{ name }}-{{ version }}.tar.gz
  url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
  sha256: {{ sha256 }}

build:
  number: 0
  script: {{ PYTHON }} -m pip install -vvv .

requirements:
  host:
    - python
    - pip
    - setuptools_scm
  run:
    - python

about:
  home: https://github.com/tqdm/tqdm
  license: MPL-2.0 AND MIT
  license_file: LICENCE
  summary: 'A Fast, Extensible Progress Bar for Python and CLI'
```

I started running into this on Friday on this repository's GitHub Actions: https://github.com/anaconda-distribution/percy/actions

I've determined locally this morning that if a `conda-build` operation runs a `pytest` with warnings, it will return with `1`. Pytest at `7.4` seems to return 0. The build says it succeeds, but still returns with `1` as an error code.

```
...
INFO :: The inputs making up the hashes for the built packages are as follows:
{
  "percy-0.1.5-py_0.tar.bz2": {
    "recipe": {}
  }
}


####################################################################################
Resource usage summary:

Total time: 0:00:41.4
CPU usage: sys=0:00:00.4, user=0:00:03.5
Maximum memory usage observed: 282.9M
Total disk usage observed (not including envs): 1.8K
Error: Process completed with exit code 1.
```

What makes matters worse (and how I ran into this in `percy`) is if you don't specify the Python version and get set to `3.12` you can hit new `pytest` deprecation warnings that you would not have seen in your previous version of Python.

--- a/conda_build/cli/main_build.py
+++ b/conda_build/cli/main_build.py
@@ -11,7 +11,7 @@ from glob import glob
 from itertools import chain
 from os.path import abspath, expanduser, expandvars
 from pathlib import Path
-from typing import Literal, Sequence, overload
+from typing import Sequence
 
 from conda.auxlib.ish import dals
 from conda.common.io import dashlist
@@ -522,20 +522,7 @@ def check_action(recipe, config):
     return api.check(recipe, config=config)
 
 
-@overload
-def execute(args: Sequence[str] | None, return_outputs: Literal[True]) -> list[str]:
-    ...
-
-
-@overload
-def execute(args: Sequence[str] | None, return_outputs: Literal[False]) -> int:
-    ...
-
-
-def execute(
-    args: Sequence[str] | None = None,
-    return_outputs: bool = False,
-) -> list[str] | int:
+def execute(args: Sequence[str] | None = None) -> int:
     _, parsed = parse_args(args)
     config = get_or_merge_config(None, **parsed.__dict__)
     build.check_external()
@@ -548,21 +535,20 @@ def execute(
 
     if "purge" in parsed.recipe:
         build.clean_build(config)
-        return []
+        return 0
 
     if "purge-all" in parsed.recipe:
         build.clean_build(config)
         config.clean_pkgs()
-        return []
+        return 0
 
-    outputs: list[str] = []
     if parsed.output:
         config.verbose = False
         config.quiet = True
         config.debug = False
-        outputs = [output_action(recipe, config) for recipe in parsed.recipe]
+        for recipe in parsed.recipe:
+            output_action(recipe, config)
     elif parsed.test:
-        outputs = []
         failed_recipes = []
         recipes = chain.from_iterable(
             glob(abspath(recipe), recursive=True) if "*" in recipe else [recipe]
@@ -584,11 +570,13 @@ def execute(
         else:
             print("All tests passed")
     elif parsed.source:
-        outputs = [source_action(recipe, config) for recipe in parsed.recipe]
+        for recipe in parsed.recipe:
+            source_action(recipe, config)
     elif parsed.check:
-        outputs = [check_action(recipe, config) for recipe in parsed.recipe]
+        for recipe in parsed.recipe:
+            check_action(recipe, config)
     else:
-        outputs = api.build(
+        api.build(
             parsed.recipe,
             post=parsed.post,
             test_run_post=parsed.test_run_post,
@@ -604,4 +592,4 @@ def execute(
     if not parsed.output and len(utils.get_build_folders(config.croot)) > 0:
         build.print_build_intermediate_warning(config)
 
-    return outputs if return_outputs else 0
+    return 0

--- a/conda_build/cli/main_build.py
+++ b/conda_build/cli/main_build.py
@@ -11,7 +11,7 @@ from glob import glob
 from itertools import chain
 from os.path import abspath, expanduser, expandvars
 from pathlib import Path
-from typing import Sequence
+from typing import Literal, Sequence, overload
 
 from conda.auxlib.ish import dals
 from conda.common.io import dashlist
@@ -522,7 +522,19 @@ def check_action(recipe, config):
     return api.check(recipe, config=config)
 
 
-def execute(args: Sequence[str] | None = None):
+@overload
+def execute(args: Sequence[str] | None, return_outputs: Literal[True]) -> list[str]:
+    ...
+
+
+@overload
+def execute(args: Sequence[str] | None, return_outputs: Literal[False]) -> int:
+    ...
+
+
+def execute(
+    args: Sequence[str] | None = None, return_outputs: bool = True
+) -> list[str] | int:
     _, parsed = parse_args(args)
     config = get_or_merge_config(None, **parsed.__dict__)
     build.check_external()
@@ -535,14 +547,14 @@ def execute(args: Sequence[str] | None = None):
 
     if "purge" in parsed.recipe:
         build.clean_build(config)
-        return
+        return []
 
     if "purge-all" in parsed.recipe:
         build.clean_build(config)
         config.clean_pkgs()
-        return
+        return []
 
-    outputs = None
+    outputs: list[str] = []
     if parsed.output:
         config.verbose = False
         config.quiet = True
@@ -590,4 +602,5 @@ def execute(args: Sequence[str] | None = None):
 
     if not parsed.output and len(utils.get_build_folders(config.croot)) > 0:
         build.print_build_intermediate_warning(config)
-    return outputs
+
+    return outputs if return_outputs else 0

--- a/conda_build/cli/main_build.py
+++ b/conda_build/cli/main_build.py
@@ -548,7 +548,9 @@ def execute(args: Sequence[str] | None = None) -> int:
         config.debug = False
         for recipe in parsed.recipe:
             output_action(recipe, config)
-    elif parsed.test:
+        return 0
+
+    if parsed.test:
         failed_recipes = []
         recipes = chain.from_iterable(
             glob(abspath(recipe), recursive=True) if "*" in recipe else [recipe]
@@ -589,7 +591,7 @@ def execute(args: Sequence[str] | None = None) -> int:
             cache_dir=parsed.cache_dir,
         )
 
-    if not parsed.output and len(utils.get_build_folders(config.croot)) > 0:
+    if utils.get_build_folders(config.croot):
         build.print_build_intermediate_warning(config)
 
     return 0

--- a/conda_build/cli/main_build.py
+++ b/conda_build/cli/main_build.py
@@ -533,7 +533,8 @@ def execute(args: Sequence[str] | None, return_outputs: Literal[False]) -> int:
 
 
 def execute(
-    args: Sequence[str] | None = None, return_outputs: bool = True
+    args: Sequence[str] | None = None,
+    return_outputs: bool = False,
 ) -> list[str] | int:
     _, parsed = parse_args(args)
     config = get_or_merge_config(None, **parsed.__dict__)

--- a/conda_build/cli/main_convert.py
+++ b/conda_build/cli/main_convert.py
@@ -120,7 +120,7 @@ all.""",
     return parser, parser.parse_args(args)
 
 
-def execute(args: Sequence[str] | None = None):
+def execute(args: Sequence[str] | None = None) -> int:
     _, parsed = parse_args(args)
     files = parsed.files
     del parsed.__dict__["files"]
@@ -128,3 +128,5 @@ def execute(args: Sequence[str] | None = None):
     for f in files:
         f = abspath(expanduser(f))
         api.convert(f, **parsed.__dict__)
+
+    return 0

--- a/conda_build/cli/main_debug.py
+++ b/conda_build/cli/main_debug.py
@@ -88,7 +88,7 @@ Set up environments and activation scripts to debug your build or test phase.
     return p
 
 
-def execute(args: Sequence[str] | None = None):
+def execute(args: Sequence[str] | None = None) -> int:
     parser = get_parser()
     parsed = parser.parse_args(args)
 
@@ -118,3 +118,5 @@ def execute(args: Sequence[str] | None = None):
             f"Error: conda-debug encountered the following error:\n{e}", file=sys.stderr
         )
         sys.exit(1)
+
+    return 0

--- a/conda_build/cli/main_develop.py
+++ b/conda_build/cli/main_develop.py
@@ -76,7 +76,7 @@ This works by creating a conda.pth file in site-packages.""",
     return parser, parser.parse_args(args)
 
 
-def execute(args: Sequence[str] | None = None):
+def execute(args: Sequence[str] | None = None) -> int:
     _, parsed = parse_args(args)
     prefix = determine_target_prefix(context, parsed)
     api.develop(
@@ -87,3 +87,5 @@ def execute(args: Sequence[str] | None = None):
         clean=parsed.clean,
         uninstall=parsed.uninstall,
     )
+
+    return 0

--- a/conda_build/cli/main_inspect.py
+++ b/conda_build/cli/main_inspect.py
@@ -184,7 +184,7 @@ Tools for investigating conda channels.
     return parser, parser.parse_args(args)
 
 
-def execute(args: Sequence[str] | None = None):
+def execute(args: Sequence[str] | None = None) -> int:
     parser, parsed = parse_args(args)
 
     if not parsed.subcommand:
@@ -221,3 +221,5 @@ def execute(args: Sequence[str] | None = None):
         pprint(api.inspect_hash_inputs(parsed.packages))
     else:
         parser.error(f"Unrecognized subcommand: {parsed.subcommand}.")
+
+    return 0

--- a/conda_build/cli/main_metapackage.py
+++ b/conda_build/cli/main_metapackage.py
@@ -108,7 +108,9 @@ command line with the conda metapackage command.
     return parser, parser.parse_args(args)
 
 
-def execute(args: Sequence[str] | None = None):
+def execute(args: Sequence[str] | None = None) -> int:
     _, args = parse_args(args)
     channel_urls = args.__dict__.get("channel") or args.__dict__.get("channels") or ()
     api.create_metapackage(channel_urls=channel_urls, **args.__dict__)
+
+    return 0

--- a/conda_build/cli/main_render.py
+++ b/conda_build/cli/main_render.py
@@ -14,6 +14,7 @@ from yaml.parser import ParserError
 from .. import __version__, api
 from ..conda_interface import ArgumentParser, add_parser_channels, cc_conda_build
 from ..config import get_channel_urls, get_or_merge_config
+from ..deprecations import deprecated
 from ..utils import LoggingContext
 from ..variants import get_package_variants, set_language_env_vars
 
@@ -189,7 +190,8 @@ def parse_args(args: Sequence[str] | None) -> tuple[ArgumentParser, Namespace]:
     return parser, parser.parse_args(args)
 
 
-def execute(args: Sequence[str] | None = None, print_results: bool = True):
+@deprecated.argument("24.1.1", "24.3.0", "print_results")
+def execute(args: Sequence[str] | None = None) -> int:
     _, parsed = parse_args(args)
 
     config = get_or_merge_config(None, **parsed.__dict__)
@@ -221,24 +223,23 @@ def execute(args: Sequence[str] | None = None, print_results: bool = True):
             f"Only one will be written to the file you specified ({parsed.file})."
         )
 
-    if print_results:
-        if parsed.output:
-            with LoggingContext(logging.CRITICAL + 1):
-                paths = api.get_output_file_paths(metadata_tuples, config=config)
-                print("\n".join(sorted(paths)))
-            if parsed.file:
-                m = metadata_tuples[-1][0]
-                api.output_yaml(m, parsed.file, suppress_outputs=True)
-        else:
-            logging.basicConfig(level=logging.INFO)
-            for m, _, _ in metadata_tuples:
-                print("--------------")
-                print("Hash contents:")
-                print("--------------")
-                pprint(m.get_hash_contents())
-                print("----------")
-                print("meta.yaml:")
-                print("----------")
-                print(api.output_yaml(m, parsed.file, suppress_outputs=True))
+    if parsed.output:
+        with LoggingContext(logging.CRITICAL + 1):
+            paths = api.get_output_file_paths(metadata_tuples, config=config)
+            print("\n".join(sorted(paths)))
+        if parsed.file:
+            m = metadata_tuples[-1][0]
+            api.output_yaml(m, parsed.file, suppress_outputs=True)
     else:
-        return metadata_tuples
+        logging.basicConfig(level=logging.INFO)
+        for m, _, _ in metadata_tuples:
+            print("--------------")
+            print("Hash contents:")
+            print("--------------")
+            pprint(m.get_hash_contents())
+            print("----------")
+            print("meta.yaml:")
+            print("----------")
+            print(api.output_yaml(m, parsed.file, suppress_outputs=True))
+
+    return 0

--- a/conda_build/cli/main_skeleton.py
+++ b/conda_build/cli/main_skeleton.py
@@ -46,7 +46,7 @@ options available.
     return parser, parser.parse_args(args)
 
 
-def execute(args: Sequence[str] | None = None):
+def execute(args: Sequence[str] | None = None) -> int:
     parser, parsed = parse_args(args)
     config = Config(**parsed.__dict__)
 
@@ -62,3 +62,5 @@ def execute(args: Sequence[str] | None = None):
         version=parsed.version,
         config=config,
     )
+
+    return 0

--- a/conda_build/plugin.py
+++ b/conda_build/plugin.py
@@ -1,55 +1,59 @@
 # Copyright (C) 2014 Anaconda, Inc
 # SPDX-License-Identifier: BSD-3-Clause
+from __future__ import annotations
+
+from typing import Sequence
+
 import conda.plugins
 
 
 # lazy-import to avoid nasty import-time side effects when not using conda-build
-def build(*args, **kwargs):
+def build(args: Sequence[str]) -> int:
     from .cli.main_build import execute
 
-    execute(*args, **kwargs)
+    return execute(args, return_outputs=False)
 
 
-def convert(*args, **kwargs):
+def convert(args: Sequence[str]) -> int:
     from .cli.main_convert import execute
 
-    execute(*args, **kwargs)
+    return execute(args)
 
 
-def debug(*args, **kwargs):
+def debug(args: Sequence[str]) -> int:
     from .cli.main_debug import execute
 
-    execute(*args, **kwargs)
+    return execute(args)
 
 
-def develop(*args, **kwargs):
+def develop(args: Sequence[str]) -> int:
     from .cli.main_develop import execute
 
-    execute(*args, **kwargs)
+    return execute(args)
 
 
-def inspect(*args, **kwargs):
+def inspect(args: Sequence[str]) -> int:
     from .cli.main_inspect import execute
 
-    execute(*args, **kwargs)
+    return execute(args)
 
 
-def metapackage(*args, **kwargs):
+def metapackage(args: Sequence[str]) -> int:
     from .cli.main_metapackage import execute
 
-    execute(*args, **kwargs)
+    return execute(args)
 
 
-def render(*args, **kwargs):
+def render(args: Sequence[str]) -> int:
     from .cli.main_render import execute
 
-    execute(*args, **kwargs)
+    return execute(args)
 
 
-def skeleton(*args, **kwargs):
+def skeleton(args: Sequence[str]) -> int:
     from .cli.main_skeleton import execute
 
-    execute(*args, **kwargs)
+    return execute(args)
 
 
 @conda.plugins.hookimpl

--- a/conda_build/plugin.py
+++ b/conda_build/plugin.py
@@ -11,7 +11,7 @@ import conda.plugins
 def build(args: Sequence[str]) -> int:
     from .cli.main_build import execute
 
-    return execute(args, return_outputs=False)
+    return execute(args)
 
 
 def convert(args: Sequence[str]) -> int:

--- a/conda_build/utils.py
+++ b/conda_build/utils.py
@@ -990,7 +990,7 @@ def get_site_packages(prefix, py_ver):
     return os.path.join(get_stdlib_dir(prefix, py_ver), "site-packages")
 
 
-def get_build_folders(croot):
+def get_build_folders(croot: str | os.PathLike | Path) -> list[str]:
     # remember, glob is not a regex.
     return glob(os.path.join(croot, "*" + "[0-9]" * 10 + "*"), recursive=True)
 

--- a/news/5169-fix-nonzero-exitcode
+++ b/news/5169-fix-nonzero-exitcode
@@ -1,0 +1,19 @@
+### Enhancements
+
+* <news item>
+
+### Bug fixes
+
+* Fix nonzero exitcode on success. (#5167 via #5169)
+
+### Deprecations
+
+* <news item>
+
+### Docs
+
+* <news item>
+
+### Other
+
+* <news item>

--- a/tests/cli/test_main_build.py
+++ b/tests/cli/test_main_build.py
@@ -130,15 +130,23 @@ def test_build_output_build_path_multiple_recipes(
     assert output.rstrip().splitlines() == test_paths, error
 
 
-def test_slash_in_recipe_arg_keeps_build_id(testing_workdir, testing_config):
+def test_slash_in_recipe_arg_keeps_build_id(
+    testing_workdir: str, testing_config: Config
+):
     args = [
         os.path.join(metadata_dir, "has_prefix_files"),
         "--croot",
         testing_config.croot,
         "--no-anaconda-upload",
     ]
-    outputs = main_build.execute(args, return_outputs=True)
-    data = package_has_file(outputs[0], "binary-has-prefix", refresh_mode="forced")
+    main_build.execute(args)
+
+    output = os.path.join(
+        testing_config.croot,
+        testing_config.host_subdir,
+        "conda-build-test-has-prefix-files-1.0-0.tar.bz2",
+    )
+    data = package_has_file(output, "binary-has-prefix", refresh_mode="forced")
     assert data
     if hasattr(data, "decode"):
         data = data.decode("UTF-8")
@@ -157,7 +165,7 @@ def test_build_long_test_prefix_default_enabled(mocker, testing_workdir):
         main_build.execute(args)
 
 
-def test_build_no_build_id(testing_workdir, testing_config):
+def test_build_no_build_id(testing_workdir: str, testing_config: Config):
     args = [
         os.path.join(metadata_dir, "has_prefix_files"),
         "--no-build-id",
@@ -166,8 +174,14 @@ def test_build_no_build_id(testing_workdir, testing_config):
         "--no-activate",
         "--no-anaconda-upload",
     ]
-    outputs = main_build.execute(args, return_outputs=True)
-    data = package_has_file(outputs[0], "binary-has-prefix", refresh_mode="forced")
+    main_build.execute(args)
+
+    output = os.path.join(
+        testing_config.croot,
+        testing_config.host_subdir,
+        "conda-build-test-has-prefix-files-1.0-0.tar.bz2",
+    )
+    data = package_has_file(output, "binary-has-prefix", refresh_mode="forced")
     assert data
     if hasattr(data, "decode"):
         data = data.decode("UTF-8")
@@ -191,7 +205,7 @@ def test_build_multiple_recipes(testing_metadata, testing_workdir, testing_confi
     main_build.execute(args)
 
 
-def test_build_output_folder(testing_workdir: str, testing_metadata):
+def test_build_output_folder(testing_workdir: str, testing_metadata: MetaData):
     api.output_yaml(testing_metadata, "meta.yaml")
 
     out = Path(testing_workdir, "out")
@@ -207,9 +221,10 @@ def test_build_output_folder(testing_workdir: str, testing_metadata):
         "--output-folder",
         str(out),
     ]
-    output = main_build.execute(args, return_outputs=True)[0]
+    main_build.execute(args)
+
     assert (
-        out / testing_metadata.config.host_subdir / os.path.basename(output)
+        out / testing_metadata.config.host_subdir / testing_metadata.pkg_fn()
     ).is_file()
 
 
@@ -375,38 +390,53 @@ def test_activate_scripts_not_included(testing_workdir):
         assert not package_has_file(out, f)
 
 
-def test_relative_path_croot(conda_build_test_recipe_envvar: str):
+def test_relative_path_croot(
+    conda_build_test_recipe_envvar: str, testing_config: Config
+):
     # this tries to build a package while specifying the croot with a relative path:
     # conda-build --no-test --croot ./relative/path
+    empty_sections = Path(metadata_dir, "empty_with_build_script")
+    croot = Path(".", "relative", "path")
 
-    empty_sections = os.path.join(metadata_dir, "empty_with_build_script")
-    croot_rel = os.path.join(".", "relative", "path")
-    args = ["--no-anaconda-upload", "--croot", croot_rel, empty_sections]
-    outputfile = main_build.execute(args, return_outputs=True)
+    args = ["--no-anaconda-upload", f"--croot={croot}", str(empty_sections)]
+    main_build.execute(args)
 
-    assert len(outputfile) == 1
-    assert os.path.isfile(outputfile[0])
+    assert len(list(croot.glob("**/*.tar.bz2"))) == 1
+    assert (
+        croot / testing_config.subdir / "empty_with_build_script-0.0-0.tar.bz2"
+    ).is_file()
 
 
-def test_relative_path_test_artifact(conda_build_test_recipe_envvar: str):
+def test_relative_path_test_artifact(
+    conda_build_test_recipe_envvar: str, testing_config: Config
+):
     # this test builds a package into (cwd)/relative/path and then calls:
     # conda-build --test ./relative/path/{platform}/{artifact}.tar.bz2
-
-    empty_sections = os.path.join(metadata_dir, "empty_with_build_script")
-    croot_rel = os.path.join(".", "relative", "path")
-    croot_abs = os.path.abspath(os.path.normpath(croot_rel))
+    empty_sections = Path(metadata_dir, "empty_with_build_script")
+    croot_rel = Path(".", "relative", "path")
+    croot_abs = croot_rel.resolve()
 
     # build the package
-    args = ["--no-anaconda-upload", "--no-test", "--croot", croot_abs, empty_sections]
-    output_file_abs = main_build.execute(args, return_outputs=True)
-    assert len(output_file_abs) == 1
+    args = [
+        "--no-anaconda-upload",
+        "--no-test",
+        f"--croot={croot_abs}",
+        str(empty_sections),
+    ]
+    main_build.execute(args)
 
-    output_file_rel = os.path.join(
-        croot_rel, os.path.relpath(output_file_abs[0], croot_abs)
-    )
+    assert len(list(croot_abs.glob("**/*.tar.bz2"))) == 1
 
     # run the test stage with relative path
-    args = ["--no-anaconda-upload", "--test", output_file_rel]
+    args = [
+        "--no-anaconda-upload",
+        "--test",
+        os.path.join(
+            croot_rel,
+            testing_config.subdir,
+            "empty_with_build_script-0.0-0.tar.bz2",
+        ),
+    ]
     main_build.execute(args)
 
 
@@ -414,17 +444,28 @@ def test_relative_path_test_recipe(conda_build_test_recipe_envvar: str):
     # this test builds a package into (cwd)/relative/path and then calls:
     # conda-build --test --croot ./relative/path/ /abs/path/to/recipe
 
-    empty_sections = os.path.join(metadata_dir, "empty_with_build_script")
-    croot_rel = os.path.join(".", "relative", "path")
-    croot_abs = os.path.abspath(os.path.normpath(croot_rel))
+    empty_sections = Path(metadata_dir, "empty_with_build_script")
+    croot_rel = Path(".", "relative", "path")
+    croot_abs = croot_rel.resolve()
 
     # build the package
-    args = ["--no-anaconda-upload", "--no-test", "--croot", croot_abs, empty_sections]
-    output_file_abs = main_build.execute(args, return_outputs=True)
-    assert len(output_file_abs) == 1
+    args = [
+        "--no-anaconda-upload",
+        "--no-test",
+        f"--croot={croot_abs}",
+        str(empty_sections),
+    ]
+    main_build.execute(args)
+
+    assert len(list(croot_abs.glob("**/*.tar.bz2"))) == 1
 
     # run the test stage with relative croot
-    args = ["--no-anaconda-upload", "--test", "--croot", croot_rel, empty_sections]
+    args = [
+        "--no-anaconda-upload",
+        "--test",
+        f"--croot={croot_rel}",
+        str(empty_sections),
+    ]
     main_build.execute(args)
 
 

--- a/tests/cli/test_main_build.py
+++ b/tests/cli/test_main_build.py
@@ -137,7 +137,7 @@ def test_slash_in_recipe_arg_keeps_build_id(testing_workdir, testing_config):
         testing_config.croot,
         "--no-anaconda-upload",
     ]
-    outputs = main_build.execute(args)
+    outputs = main_build.execute(args, return_outputs=True)
     data = package_has_file(outputs[0], "binary-has-prefix", refresh_mode="forced")
     assert data
     if hasattr(data, "decode"):
@@ -166,7 +166,7 @@ def test_build_no_build_id(testing_workdir, testing_config):
         "--no-activate",
         "--no-anaconda-upload",
     ]
-    outputs = main_build.execute(args)
+    outputs = main_build.execute(args, return_outputs=True)
     data = package_has_file(outputs[0], "binary-has-prefix", refresh_mode="forced")
     assert data
     if hasattr(data, "decode"):
@@ -207,7 +207,7 @@ def test_build_output_folder(testing_workdir: str, testing_metadata):
         "--output-folder",
         str(out),
     ]
-    output = main_build.execute(args)[0]
+    output = main_build.execute(args, return_outputs=True)[0]
     assert (
         out / testing_metadata.config.host_subdir / os.path.basename(output)
     ).is_file()
@@ -382,7 +382,7 @@ def test_relative_path_croot(conda_build_test_recipe_envvar: str):
     empty_sections = os.path.join(metadata_dir, "empty_with_build_script")
     croot_rel = os.path.join(".", "relative", "path")
     args = ["--no-anaconda-upload", "--croot", croot_rel, empty_sections]
-    outputfile = main_build.execute(args)
+    outputfile = main_build.execute(args, return_outputs=True)
 
     assert len(outputfile) == 1
     assert os.path.isfile(outputfile[0])
@@ -398,7 +398,7 @@ def test_relative_path_test_artifact(conda_build_test_recipe_envvar: str):
 
     # build the package
     args = ["--no-anaconda-upload", "--no-test", "--croot", croot_abs, empty_sections]
-    output_file_abs = main_build.execute(args)
+    output_file_abs = main_build.execute(args, return_outputs=True)
     assert len(output_file_abs) == 1
 
     output_file_rel = os.path.join(
@@ -420,7 +420,7 @@ def test_relative_path_test_recipe(conda_build_test_recipe_envvar: str):
 
     # build the package
     args = ["--no-anaconda-upload", "--no-test", "--croot", croot_abs, empty_sections]
-    output_file_abs = main_build.execute(args)
+    output_file_abs = main_build.execute(args, return_outputs=True)
     assert len(output_file_abs) == 1
 
     # run the test stage with relative croot

